infrastructure/src/main/java/org/corfudb/infrastructure/CompactorLeaderServices.java

```diff
@@ -0,0 +1,497 @@
+
+            this.checkpointTable = this.corfuStore.openTable(CORFU_SYSTEM_NAMESPACE,
+                    DistributedCompactor.CHECKPOINT,
+                    StringKey.class,
+                    RpcCommon.TokenMsg.class,
+                    null,
+                    TableOptions.fromProtoSchema(RpcCommon.TokenMsg.class));
+
+        } catch (Exception e) {
+            syslog.error("Caught an exception while opening Compaction management tables ", e);
+        }
+    }
+
+    /**
+     * The leader initiates the Distributed checkpointer
+     * Trims the log till 'trimtoken' saved in the checkpoint table
+     * Mark the start of the compaction cycle and populate CheckpointStatusTable with all the
+     * tables in the registry.
+     *
+     * @return
+     */
+    @VisibleForTesting
+    public boolean trimAndTriggerDistributedCheckpointing() {
+        syslog.info("=============Initiating Distributed Checkpointing============");
+        if (!isLeader) {
+            return false;
+        }
+        trimLog();
+        if (DistributedCompactor.isCheckpointFrozen(corfuStore, checkpointTable)) {
+            syslog.warn("Will not start checkpointing since checkpointing has been frozen");
+            return false;
+        }
+
+        try (TxnContext txn = corfuStore.txn(CORFU_SYSTEM_NAMESPACE)) {
+            CorfuStoreEntry compactionManagerRecord = txn.getRecord(
+                    DistributedCompactor.COMPACTION_MANAGER_TABLE_NAME, DistributedCompactor.COMPACTION_MANAGER_KEY);
+            CheckpointingStatus managerStatus = (CheckpointingStatus) compactionManagerRecord.getPayload();
+
+            if (managerStatus != null && (managerStatus.getStatus() == StatusType.STARTED ||
+                    managerStatus.getStatus() == StatusType.STARTED_ALL)) {
+                return false;
+            }
+
+            List<TableName> tableNames = new ArrayList<>(corfuStore.listTables(null));
+            CheckpointingStatus idleStatus = buildCheckpointStatus(StatusType.IDLE);
+
+            TransactionalContext.getCurrentContext().setPriorityLevel(CorfuMessage.PriorityLevel.HIGH);
+            txn.clear(checkpointingStatusTable);
+            txn.clear(activeCheckpointsTable);
+            //Populate CheckpointingStatusTable
+            for (TableName table : tableNames) {
+                txn.putRecord(checkpointingStatusTable, table, idleStatus, null);
+            }
+
+            // Also record the minToken as the earliest token BEFORE checkpointing is initiated
+            // This is the safest point to trim at since all data up to this point will surely
+            // be included in the upcoming checkpoint cycle
+            long minAddressBeforeCycleStarts = corfuRuntime.getAddressSpaceView().getLogTail();
+            txn.putRecord(checkpointTable, DistributedCompactor.CHECKPOINT_KEY,
+                    RpcCommon.TokenMsg.newBuilder()
+                            .setEpoch(this.epoch)
```

Comment:
It looks like this epoch is always 0. Is this intended?

Reply:
actually epoch can be removed from here. trim ignores it anyway